effect/README.md

```diff
@@ -0,0 +1,294 @@
+
+## Brackets
+
+Brackets are a built-in primitive that let you safely acquire and release resources.
+
+Brackets are used for a similar purpose as try/catch/finally, only brackets work with synchronous and asynchronous actions, work seamlessly with fiber interruption, and are built on a different error model that ensures no errors are ever swallowed.
+
+Brackets consist of an *acquire* action, a *utilize* action (which uses the acquired resource), and a *release* action.
+
+The release action is guaranteed to be executed by the runtime system, even if the utilize action throws an exception or the executing fiber is interrupted.
+
+```scala
+openFile("data.json").bracket(closeFile(_)) { file =>
+  for {
+    data    <- decodeData(file)
+    grouped <- groupData(data)
+  } yield grouped
+}
+```
+
+Brackets have compositional semantics, so if a bracket is nested inside another bracket, and the outer bracket acquires a resource, then the outer bracket's release will always be called, even if, for example, the inner bracket's release fails.
+
+A helper method called `ensuring` provides a simpler analogue of `finally`:
+
+```scala
+val composite = action1.ensuring(cleanupAction)
+```
+
+## Fibers
+
+To perform an action without blocking the current process, you can use fibers, which are a lightweight mechanism for concurrency.
+
+You can `fork` any `IO[A]` to immediately yield an `IO[Fiber[A]]`. The provided `Fiber` can be used to `join` the fiber, which will resume on production of the fiber's value, or to `interrupt` the fiber with some exception.
+
+```scala
+val analyzed =
+  for {
+    fiber1   <- analyzeData(data).fork  // IO[Analysis]
+    fiber2   <- validateData(data).fork // IO[Boolean]
+    ... // Do other stuff
+    valid    <- fiber2.join
+    _        <- if (!valid) fiber1.interrupt(DataValidationError(data))
+                else IO.unit
+    analyzed <- fiber1.join
+  } yield analyzed
+```
+
+On the JVM, fibers will use threads, but will not consume *unlimited* threads. Instead, fibers yield cooperatively during periods of high-contention.
+
+```scala
+def fib(n: Int): IO[Int] =
+  if (n <= 1) IO.point(1)
+  else for {
+    fiber1 <- fib(n - 2).fork
+    fiber2 <- fib(n - 1).fork
+    v2     <- fiber2.join
+    v1     <- fiber1.join
+  } yield v1 + v2
+```
+
+Interrupting a fiber returns an action that resumes when the fiber has completed or has been interrupted and all its finalizers have been run. These semantics allow construction of programs that do not leak resources.
+
+A more powerful variant of `fork`, called `fork0`, allows specification of a handler that will be passed any uncaught errors from the forked fiber, including all uncaught errors that occur in finalizers. If this handler is not specified, then the handler of the parent fiber will be used, recursively, up to the root handler, which can be specified in `RTS`.
```

Comment:
could the `fork0` have a better name?

Reply:
I'd like to use `fork'` but it's not allowed in Scala. 😄 

I'm open to suggestions. Generally I use `0` variants to define the most powerful version of a function (e.g. `bracket0`), from which all other variants are defined.

Reply:
Or just overload `fork`?

Reply:
We should have a policy on this. Personally, I think it's OK to use overloading _only_ for methods (never for type class functions or top-level functions / syntax).